server/hotrod/src/main/java/org/infinispan/server/hotrod/Authentication.java

```diff
@@ -86,74 +75,105 @@ public void auth(HotRodHeader header, String mech, byte[] response) throws Privi
             server.accessLogging().logException(future, (AccessLoggingHeader) header, cause.toString(), responseBytes);
          }
       } else {
-         if (saslServer == null) {
-            ServerAuthenticationProvider sap = authenticationConfig.serverAuthenticationProvider();
-            callbackHandler = sap.getCallbackHandler(mech, authenticationConfig.mechProperties());
-            final SaslServerFactory ssf;
-            if ("EXTERNAL".equals(mech)) {
-               SslHandler sslHandler = channel.pipeline().get(SslHandler.class);
-               try {
-                  if (sslHandler != null)
-                     ssf = new ExternalSaslServerFactory(sslHandler.engine().getSession().getPeerPrincipal());
-                  else
-                     throw log.externalMechNotAllowedWithoutSSLClientCert();
-               } catch (SSLPeerUnverifiedException e) {
+         executor.execute(() -> authInternal(header, mech, response));
+      }
+   }
+
+   private void authInternal(HotRodHeader header, String mech, byte[] response) {
+      if (saslServer == null) {
+         ServerAuthenticationProvider sap = authenticationConfig.serverAuthenticationProvider();
+         callbackHandler = sap.getCallbackHandler(mech, authenticationConfig.mechProperties());
+         final SaslServerFactory ssf;
+         if ("EXTERNAL".equals(mech)) {
+            SslHandler sslHandler = channel.pipeline().get(SslHandler.class);
+            try {
+               if (sslHandler != null)
+                  ssf = new ExternalSaslServerFactory(sslHandler.engine().getSession().getPeerPrincipal());
+               else
                   throw log.externalMechNotAllowedWithoutSSLClientCert();
-               }
-            } else {
-               ssf = server.getSaslServerFactory(mech);
+            } catch (SSLPeerUnverifiedException e) {
+               SecurityException securityException = log.externalMechNotAllowedWithoutSSLClientCert();
+               writeException(header, securityException);
+               throw securityException;
             }
-            if (authenticationConfig.serverSubject() != null) {
+         } else {
+            ssf = server.getSaslServerFactory(mech);
+         }
+         if (authenticationConfig.serverSubject() != null) {
+            try {
                saslServer = Subject.doAs(authenticationConfig.serverSubject(), (PrivilegedExceptionAction<SaslServer>) () ->
                      ssf.createSaslServer(mech, "hotrod", authenticationConfig.serverName(),
                            authenticationConfig.mechProperties(), callbackHandler));
-            } else {
+            } catch (PrivilegedActionException e) {
+               throw new SecurityException(writeException(header, e.getCause()));
```

Comment:
Why are you throwing that here? Can't you just log the PAE, writeException and `return`?